### PR TITLE
Feature/validatation tests

### PR DESCRIPTION
**Pull request by santhoshMalay**
Added validation scenarios for sign-up.
Also added the unit tests for various scenarios of log-in and sign-up end points.

### CODE REVIEW

File: internal/delivery/http/v1/auth.go [hunk 1]
```diff
@@ -24,8 +24,8 @@ func (h *Handler) initAuthRoutes(api *gin.RouterGroup) {
 // @Accept  json
 // @Produce  json
 // @Param input body service.SignupUserInput true "New user signup details"
```
Comment (zhuravlev-pe) [inline internal/delivery/http/v1/auth.go:26]: Swagger definitions below are incorrect:
- success does not return an object
- there is no 404 response, but there should be a 400 response for validation failure
- I recommend using utils.ValidationError as return type for 400 response, as this is what h.handleServiceError() writes to output on a validation error

In addition to that: 
- I recommend changing the success code to 204 (http.StatusNoContent) and change the swagger and the method code accordingly.
- there is a new method h.parseRequestBody() (in helpers.go) which I recommend using here to reduce code duplication

File: internal/delivery/http/v1/auth.go [hunk 2]
```diff
@@ -66,13 +61,7 @@ func (h *Handler) userLogin(ctx *gin.Context) {
...
 	if err != nil {
-		if err == service.ErrInvalidCredentials {
-			utils.ErrorResponse(ctx, http.StatusBadRequest, err)
-			return
-		}
-
-		utils.ErrorResponse(ctx, http.StatusInternalServerError, err)
-		return
+		h.handleServiceError(ctx, err)
 	}
```
Comment (zhuravlev-pe) [inline internal/delivery/http/v1/auth.go:65]: This method returns service.PostUserLoginOutput on success. Please update swagger. Also it makes sense using h.parseRequestBody() helper here too

File: internal/service/users.go [hunk 3]
```diff
@@ -55,7 +56,19 @@ func (u *usersService) UpdateUserInfo(ctx context.Context, id string, input *Upd
 	return u.repo.Update(ctx, id, &upd)
 }
 
+func (i *SignupUserInput) Validate() error {
+	return validation.ValidateStruct(i,
+		validation.Field(&i.Email, validation.Required, is.Email),
```
Comment (zhuravlev-pe) [inline internal/service/users.go:61]: Eventually it makes sense to make password complexity check configurable (via parameters in Config), but let's keep this simple implementation for now

File: internal/service/users_test.go [hunk 4]
```diff
@@ -214,3 +217,255 @@ func TestUsersService_UpdateUserInfo(t *testing.T) {
...
+			input: &SignupUserInput{
+				Email:       "JognDoe@example.com",
+				Password:    "John*123#",
+				FirstName:   "John",
+				LastName:    "Doe",
+				DisplayName: "JohnnyD",
+			},
+			setupMocks: func(ctx context.Context, mockUsers *repoMocks.MockUsers) {
+				u := &core.User{
+					Id:               "1111111",
```
Comment (zhuravlev-pe) [inline internal/service/users_test.go:253]: This test is correct, but may be a bit misleading, as emails are different. Please make the emails equal if possible

File: internal/service/users_test.go [hunk 5]
```diff
@@ -214,3 +217,255 @@ func TestUsersService_UpdateUserInfo(t *testing.T) {
...
+					LastName:         "Doe",
+					DisplayName:      "JohnnyD",
+					RegistrationDate: someDate,
+					HashedPassword:   hashedPassword,
+					Roles:            []security.Role{security.Student},
+				}
+				mockUsers.EXPECT().GetByEmail(ctx, "JognDoe@example.com").Return(loggedUser, nil).Times(1)
+			},
+			checkError: noError,
+		},
```
Comment (zhuravlev-pe) [inline internal/service/users_test.go:412]: "invalid_password"

File: internal/service/users_test.go [hunk 6]
```diff
@@ -214,3 +217,255 @@ func TestUsersService_UpdateUserInfo(t *testing.T) {
...
+				Persistent: true,
+			},
+			setupMocks: func(ctx context.Context, mockUsers *repoMocks.MockUsers) {
+				loggedUser := &core.User{
+					Id:               "1111111",
+					Email:            "doe.h@example.com",
+					FirstName:        "John",
+					LastName:         "Doe",
+					DisplayName:      "JohnnyD",
+					RegistrationDate: someDate,
```
Comment (zhuravlev-pe) [inline internal/service/users_test.go:428]: I'd recommend to be more explicit here: hash an explicitly different password (say, password + "qwerty") using bcrypt and put it here. Currently you are testing a somewhat different situation: passing invalid hashed data to bcrypt (if you look closely at the HashedPassword generated by bcrypt, you will see that it has a very specific data format). This is not what we expect when a user enters a wrong password

**Review by zhuravlev-pe — CHANGES_REQUESTED** [review]
Looking good, just a few minor changes seem to be needed :)

File: internal/service/users_test.go [hunk 7]
```diff
@@ -214,3 +217,255 @@ func TestUsersService_UpdateUserInfo(t *testing.T) {
...
+					HashedPassword:   []byte(password),
+					Roles:            []security.Role{security.Student},
+				}
```
Comment (zhuravlev-pe) [inline internal/service/users_test.go:431]: Whoops, missed this. GetByEmail() does not return ErrInvalidCredentials

File: internal/service/users_test.go [hunk 8]
```diff
@@ -214,3 +217,255 @@ func TestUsersService_UpdateUserInfo(t *testing.T) {
...
+			setupMocks: func(ctx context.Context, mockUsers *repoMocks.MockUsers) {
+				u := &core.User{
+					Id:               "1111111",
+					Email:            "doe.h@example.com",
+					FirstName:        "John",
+					LastName:         "Doe",
+					DisplayName:      "JohnnyD",
+					RegistrationDate: someDate,
+					Roles:            []security.Role{security.Student},
+				}
```
Comment (zhuravlev-pe) [inline internal/service/users_test.go:260]: Wrong return valuer here too

File: internal/service/users_test.go (hunk 4 above)
Comment (zhuravlev-pe) [reply internal/service/users_test.go:253]: Correction: wrong return value in mock makes this test invalid

**Comment by mancerayder515** [comment]
Why is there an "email" field in every testcase?

One testcase for a validation error is enough, isn't it? If we want to test how `Validate()` works should, probably, do it in a different test. Correct, if I'm wrong, please.

**Comment by zhuravlev-pe** [comment]
Yes, email fields are not used anywhere and can/should be removed
```
cases := map[string]struct {
		email      string  // <------
		input      *LoginInput
		setupMocks func(context.Context, *repoMocks.MockUsers)
		checkError func(*testing.T, error)
	}{}
```

File: internal/delivery/http/v1/auth.go (hunk 1 above)
Comment (santhoshMalay) [reply internal/delivery/http/v1/auth.go:26]: Addressed in the latest commit.

File: internal/service/users.go (hunk 3 above)
Comment (santhoshMalay) [reply internal/service/users.go:61]: Addressed in the latest commit.

File: internal/service/users_test.go (hunk 5 above)
Comment (santhoshMalay) [reply internal/service/users_test.go:412]: Addressed in the latest commit.

File: internal/service/users_test.go (hunk 7 above)
Comment (santhoshMalay) [reply internal/service/users_test.go:431]: Addressed in the latest commit.

File: internal/service/users_test.go (hunk 8 above)
Comment (santhoshMalay) [reply internal/service/users_test.go:260]: Addressed in the latest commit.

File: internal/service/users_test.go (hunk 6 above)
Comment (santhoshMalay) [reply internal/service/users_test.go:428]: Addressed in the latest commit.

File: internal/delivery/http/v1/auth.go (hunk 2 above)
Comment (santhoshMalay) [reply internal/delivery/http/v1/auth.go:65]: Addressed in the latest commit.

Comment (zhuravlev-pe) [reply internal/delivery/http/v1/auth.go:65]: The part about swagger is unresolved.

What you have:
```
// @Success 200 {object} service.LoginInput
```

In reality, method returns service.PostUserLoginOutput to the caller.

**Review by zhuravlev-pe — CHANGES_REQUESTED** [review]
Please resolve the issue with swagger comments

**Comment by zhuravlev-pe** [comment]
You did not rebuild the swagger after the last change, but let's keep it. It will be rebuild after the next merge anyway.

Please pay maximal attention to the swagger definition. API is created to be consumed by other developers working on dependent projects, both frontend and backend. The correct API definition is also important for yourself and your team, whenever you come back to this interface. Inconsistency between swagger documentation and the implementation can lead to major confusion. The API definition is the public "face" of our project and we must keep it top-notch.

We are not working on a real product, but the skill needed to keep the API definition up-to-date is very much real and needed in any actual backend work.

**Review by zhuravlev-pe — APPROVED** [review]

**Pull request merged**